code/game/objects/structures/crateshelf.dm

```diff
@@ -0,0 +1,136 @@
+#define DEFAULT_SHELF_CAPACITY 3 // Default capacity of the shelf
+#define DEFAULT_SHELF_USE_DELAY 1 SECONDS // Default interaction delay of the shelf
+#define DEFAULT_SHELF_VERTICAL_OFFSET 10 // Vertical pixel offset of shelving-related things. Set to 10 by default due to this leaving more of the crate on-screen to be clicked.
+
+/obj/structure/crate_shelf
+	name = "crate shelf"
+	desc = "It's a shelf! For storing crates!"
+	icon = 'icons/obj/objects.dmi'
+	icon_state = "shelf_base"
+	density = TRUE
+	anchored = TRUE
+	max_integrity = 50 // Not hard to break
+
+	var/capacity = DEFAULT_SHELF_CAPACITY
+	var/use_delay = DEFAULT_SHELF_USE_DELAY
+	var/list/shelf_contents
+
+/obj/structure/crate_shelf/tall
+	capacity = 12
+
+/obj/structure/crate_shelf/Initialize()
+	. = ..()
+	shelf_contents = new/list(capacity) // Initialize our shelf's contents list, this will be used later.
+	var/stack_layer // This is used to generate the sprite layering of the shelf pieces.
+	var/stack_offset // This is used to generate the vertical offset of the shelf pieces.
+	for(var/i in 1 to (capacity - 1))
+		stack_layer  = BELOW_OBJ_LAYER + (0.02 * i) - 0.01 // Make each shelf piece render above the last, but below the crate that should be on it.
+		stack_offset = DEFAULT_SHELF_VERTICAL_OFFSET * i // Make each shelf piece physically above the last.
+		overlays += image(icon = 'icons/obj/objects.dmi', icon_state = "shelf_stack", layer = stack_layer, pixel_y = stack_offset)
+	return
+
+/obj/structure/crate_shelf/Destroy()
+	var/turf/dump_turf = drop_location()
+	for(var/obj/structure/closet/crate/crate in shelf_contents)
+		crate.layer = initial(crate.layer) // Reset the crates back to default visual state
+		crate.pixel_y = initial(crate.pixel_y)
+		crate.forceMove(dump_turf)
+		step(crate, pick(GLOB.alldirs)) // Shuffle the crates around as though they've fallen down.
+		crate.SpinAnimation(rand(4,7), 1) // Spin the crates around a little as they fall. Randomness is applied so it doesn't look weird.
+		switch(pick(1, 2, 3)) // Randomly pick whether to do nothing, open the crate, or break it open.
+			if(1) // Believe it or not, this does nothing.
+			if(2) // Open the crate!
+				if(crate.open()) // Break some open, cause a little chaos.
+					crate.visible_message("<span class='warning'>[crate]'s lid falls open!</span>")
+				else // If we somehow fail to open the crate, just break it instead!
+					crate.visible_message("<span class='warning'>[crate] falls apart!")
+					crate.deconstruct()
+			if(3) // Break that crate!
+				crate.visible_message("<span class='warning'>[crate] falls apart!")
+				crate.deconstruct()
+	shelf_contents.Cut()
+	return ..()
```

Comment:
I'd move this all to `/proc/deconstruct`. The destroy for this should ideally just be `QDEL_LIST(shelf_contents)` since we don't want side effects in Destroy.